File: pythonCode/Semester.py
from LetterGrade import LetterGrade

class Semester():

    def __init__(self):
        self.__gradeLectureList = dict()
        pass

    # Creating properties for variables

    def getGradeLectureList(self):
        return self.__gradeLectureList

    def addGradeToList(self, lectureSession, letterGrade):
        self.getGradeLectureList()[lectureSession.getLecture()] = letterGrade

    def getListOfLecturesTaken(self):
        return self.__gradeLectureList

    def getCreditsTaken(self):
        self.__creditsTaken = self.__creditsTakenCalculator(self.__gradeLectureList)
        return self.__creditsTaken

    def getCreditsCompleted(self):
        self.__creditsCompleted = self.__creditsCompletedCalculator(self.__gradeLectureList)
        return self.__creditsCompleted

    def getPoints(self):
        self.__points = self.__pointsCalculator(self.__gradeLectureList)
        return self.__points

    def getYano(self):
        self.__yano = self.__points/self.__creditsTaken
        return self.__yano

    def setListOfLecturesTaken(self, listOfLecturesTaken : dict):
        self.__gradeLectureList = listOfLecturesTaken

    def addLecture(self, lecture, grade : LetterGrade):
        self.__gradeLectureList[lecture] = grade

    def __creditsTakenCalculator(self, listOfLecturesTaken : dict):
        totalCredit = 0
        for lecture in listOfLecturesTaken.keys():
            totalCredit = totalCredit + lecture.getCredit()
        return totalCredit

    def __creditsCompletedCalculator(self, listOfLecturesTaken : dict):
        totalCredit = 0
        for lecture in listOfLecturesTaken.keys():
            if listOfLecturesTaken[lecture] != LetterGrade.FD and listOfLecturesTaken[lecture] != LetterGrade.FF:
                totalCredit = totalCredit + lecture.getCredit()
        return totalCredit

    def __pointsCalculator(self, listOfLecturesTaken : dict):
        points = 0
        for lecture in listOfLecturesTaken.keys():
            points = points + (lecture.getCredit() * listOfLecturesTaken[lecture].value)
        return points

File: Iteration 2/Java Code/lecture/Semester.java
package lecture;

import java.util.HashMap;
import java.util.Map;

import Enums.LetterGrade;
import Enums.Term;
import Enums.TermYear;

public class Semester {

	private Map<Lecture, LetterGrade> listOfLecturesTaken;
	private int creditsTaken;
	private int creditsCompleted;
	private double points;
	private double yano;
	private Term term;
	private TermYear termYear;

	public Semester(Map<Lecture, LetterGrade> listOfLecturesTaken) {
		super();
		this.listOfLecturesTaken = listOfLecturesTaken;
		if (this.listOfLecturesTaken == null) {
			this.listOfLecturesTaken = new HashMap<Lecture, LetterGrade>();
		}
		this.creditsTaken = creditsTakenCalculator(this.listOfLecturesTaken);
		this.creditsCompleted = creditsCompletedCalculator(this.listOfLecturesTaken);
		this.points = pointsCalculator(this.listOfLecturesTaken);
		this.yano = points / creditsTaken;
		
	}

	// Creating get and set methods for variables
	public Map<Lecture, LetterGrade> getListOfLecturesTaken() {
		return listOfLecturesTaken;
	}

	public int getCreditsTaken() {
		return creditsTaken;
	}

	public int getCreditsCompleted() {
		return creditsCompleted;
	}

	public double getPoints() {
		return points;
	}

	public double getYano() {
		return yano;
	}

	public void setListOfLecturesTaken(Map<Lecture, LetterGrade> listOfLecturesTaken) {
		this.listOfLecturesTaken = listOfLecturesTaken;
	}
	
	public void addLecture(Lecture lecture) {
		this.listOfLecturesTaken.put(lecture, LetterGrade.None);
	}
	
	public void addLecture(Lecture lecture, LetterGrade grade) {
		this.listOfLecturesTaken.put(lecture, grade);
	}
	
	private int creditsTakenCalculator(Map<Lecture, LetterGrade> listOfLecturesTaken){
		int totalCredit = 0;
		for(Lecture lecture : listOfLecturesTaken.keySet()) {
			totalCredit = totalCredit + lecture.getCredit();
		}
		
		return totalCredit;
	}
	
	private int creditsCompletedCalculator(Map<Lecture, LetterGrade> listOfLecturesTaken){
		int totalCredit = 0;
		for(Lecture lecture : listOfLecturesTaken.keySet()) {
			if(listOfLecturesTaken.get(lecture) != LetterGrade.FD && listOfLecturesTaken.get(lecture) != LetterGrade.FF)
				totalCredit = totalCredit + lecture.getCredit();
		}
		
		return totalCredit;
	}
	
	private double pointsCalculator(Map<Lecture, LetterGrade> listOfLecturesTaken) {
		double points = 0;
		for(Lecture lecture : listOfLecturesTaken.keySet()) {
			points = points + (lecture.getCredit() * listOfLecturesTaken.get(lecture).getLetterGradeValue());
		}
		
		return points;
	}

}